Write programs for the following questions. When writing the programs, you must not use a built-in library function that directly implements the specified behavior. For example, the program for (1) must not include a call to to_i in Ruby. 

(1)	Write a program that reads a number in the quaternary representation (base-4 positional notation) and prints it in the decimal representation. For example, it reads 123 and prints 27.  

In [1]:
inNum=123
outNum=0
i=0
while inNum>0:
    outNum+=(inNum%10)*(4**i)
    inNum = int((inNum-inNum%10)/10)
    i+=1

print(outNum)

27


(2)	Suppose that symbols a, b, c, ... , h denote numbers 0, 1, 2, ... , 7, respectively. Write a program that reads a number expressed with a, b, c, ... , h in the octal representation and prints the number in the decimal representation. For example, it reads bcd and prints 83.  

In [2]:
numDic={"a":0,"b":1,"c":2,"d":3,"e":4,"f":5,"g":6,"h":7}

inStr = "bcd"
outNum = 0

for i in range(len(inStr)):
    c = inStr[i]
    tmp = numDic[c]*(8**(len(inStr)-i-1))
    outNum += tmp
    
print(outNum)


83


(3)	Write on the answer sheet decimal number 2015 in Roman numerals.  
Roman numerals use seven symbols I, V, X, L, C, D, and M. Their values are 1, 5, 10, 50, 100, 500, and 1000, respectively. They are formed as follows: 

>Numbers are formed by combining symbols and adding the values. So II is two ones, i.e. 2, and XIII is a ten and three ones, i.e. 13. There is no zero in this system, so 207, for example, is CCVII, using the symbols for two hundreds, a five and two ones. 1066 is MLXVI, one thousand, fifty and ten, a five and a one. 
>Symbols are placed from left to right in order of value, starting with the largest. However, in a few specific cases, to avoid four characters being repeated in succession (such as IIII or XXXX) these can be reduced using subtractive notation as follows:  

+	the numeral I can be placed before V and X to make 4 units (IV) and 9 units(IX)	respectively
+	X can be placed before L and C to make 40 (XL) and 90 (XC) respectively
+	C can be placed before· D and M to make 400 (CD) and 900 ( CM)

>An example using the above rules would be 1904: this is composed of 1 ( one thousand), 9 (nine hundreds), 0 (zero tens), and 4 (four units). To write the Roman numeral, each of the non-zero digits should be treated separately. Thus 1,000 = M, 900 = CM, and 4 = IV. Therefore, 1904 is MCMIV.   
>(Reference: http://en.wikipedia.org/wiki/Roman_numerals) 

The symbols have to be selected so that the number of the symbols will be the minimum. For example, IV is composed of two symbols. 

MMXV

(4)	Write a program that reads a number in Roman numerals and prints it in the decimal representation. Assume that the number is more than 0 and less than 4000.  


In [3]:
RomanNumerals ={1:"I", 5:"V",10:"X",50:"L", 100:"C", 500:"D",1000:"M"}
RomanValues ={"I":1, "V":5, "X":10, "L":50, "C":100, "D":500, "M":1000}
inStr="MMXIV"
inStr="CIL"
outNum=0
i = len(inStr)-1
while i>=0:

    if i>0 and RomanValues[inStr[i-1]]<RomanValues[inStr[i]]:
        outNum+=RomanValues[inStr[i]]
        outNum-=RomanValues[inStr[i-1]]
        i-=1
    else:
        outNum+=RomanValues[inStr[i]]
    i-=1

print(outNum)        

149



(5)	Write a program that reads a number in the decimal representation and prints it in Roman numerals. The number is more than O and less than 4000.  

In [4]:

inNum = 1904
outList=[]
outStr = ""
for k,v in sorted(RomanNumerals.items(), reverse=True):
    cnt = int(inNum/k)
    if cnt>=1:
        if cnt==4:
            pre = outList.pop()
            if pre == RomanNumerals[ k*5 ]:
                outList.append(v)
                outList.append(RomanNumerals[k*10])
            else:
                outList.append(pre)
                outList.append(v)
                outList.append(RomanNumerals[k*5])
        else:
            outList.extend([v]*cnt)
        inNum = inNum%k
outStr = ''.join(outList)

print(outStr)


MCMIV



(6)	Extend the subtraction notation of Roman numerals as follows.  
>A symbol can be placed between adjacent larger symbols a and (3 than that symbol, where $\alpha > \beta$ or $\alpha$ is blank. For example, CIL is 149 because 100 -1 + 50 = 149. IL is 49 because -1 + 50 = 49.   

Write a program that reads a number in the decimal representation and prints it in the extended Roman numerals. The number is more than 0 and less than 4000. The symbols have to be selected so that the number of the symbols will be the minimum.   


(7)Write a program that reads a number expressed in English and prints it in the decimal representation. The number is a positive integer less than 100000. For example, it reads:
>fifty four thousand three hundred twelve   

and prints 54312. The program may accept at least one English expression for each number. For example, for 1200, the program may only accept either one thousand two hundred or twelve hundred. 